crates/sui-types/src/object.rs

```diff
@@ -355,46 +358,93 @@ impl MoveObject {
 
     /// Get the total amount of SUI embedded in `self`. Intended for testing purposes
     pub fn get_total_sui(&self, layout_resolver: &mut dyn LayoutResolver) -> Result<u64, SuiError> {
-        if self.type_.is_gas_coin() {
-            // Fast path without deserialization.
-            return Ok(self.get_coin_value_unsafe());
-        }
-        // If this is a coin but not a SUI coin, the SUI balance must be 0.
-        if self.type_.is_coin() {
-            return Ok(0);
+        let balances = self.get_coin_balances(layout_resolver)?;
+        Ok(balances.get(&GAS::type_tag()).copied().unwrap_or(0))
+    }
+}
+
+// Helpers for extracting Coin<T> balances for all T
+impl MoveObject {
+    fn is_balance(s: &StructTag) -> Option<&TypeTag> {
+        (Balance::is_balance(s) && s.type_params.len() == 1).then(|| &s.type_params[0])
+    }
+
+    /// Get the total balances for all `Coin<T>` embedded in `self`.
+    pub fn get_coin_balances(
+        &self,
+        layout_resolver: &mut dyn LayoutResolver,
+    ) -> Result<BTreeMap<TypeTag, u64>, SuiError> {
+        let mut balances = BTreeMap::default();
+
+        // Fast path without deserialization.
+        if let Some(type_tag) = self.type_.coin_type_maybe() {
+            let balance = self.get_coin_value_unsafe();
```

Comment:
If `balance` is zero, we probably want to avoid updating the table--seems useful to ensure that the map's balances are all >0.